src/secret_processors.py

```diff
@@ -0,0 +1,171 @@
+        parsed_secrets_data: Parsed secrets data.
+
+    Returns:
+        dict: A dictionary containing Juju secrets.
+
+    Raises:
+        ValueError: If the Juju version does not support Juju user secrets,
+                    if a specified Juju secret is not found,
+                    or if the charm does not have permission to access the specified Juju secret.
+    """
+    charm_env = {}
+    if parsed_secrets_data.get("juju") and not JujuVersion.from_environ().has_secrets:
+        raise ValueError("Juju version does not support Juju user secrets")
+
+    juju_variables = parsed_secrets_data.get("juju", [])
+    for juju_secret in juju_variables:
+        try:
+            secret_id = juju_secret.get("secret-id")
+            key = juju_secret.get("key")
+            secret = charm.model.get_secret(id=secret_id)
+            secret_content = secret.get_content(refresh=True)
+            charm_env.update({key: secret_content[key]})
+        except SecretNotFoundError as e:
+            raise ValueError(f"Juju secret `{secret_id}` not found") from e
+        except ModelError as e:
+            raise ValueError(f"Access permission not granted to charm for secret `{secret_id}`") from e
+        except KeyError as e:
+            logger.error(f"Error parsing secrets env: {e}")
+            raise ValueError(f"Error parsing secrets env: {e}") from e
+
+    return charm_env
+
+
+def process_vault_secrets(charm, parsed_secrets_data):
+    """Process Vault secrets from the parsed secrets data.
+
+    Args:
+        charm: The charm to perform operations on.
+        parsed_secrets_data: Parsed secrets data.
+
+    Returns:
+        dict: A dictionary containing Vault secrets.
+
+    Raises:
+        ValueError: If there is no vault relation, if there is an error initializing the vault client,
+                    or if there is an error reading a vault secret.
+    """
+    # TODO (kelkawi-a): Convert to using structured config
+    charm_env = {}
+    vault_variables = parsed_secrets_data.get("vault", [])
+
+    if vault_variables and not charm.model.relations["vault"]:
+        raise ValueError("No vault relation found to fetch secrets from")
+
+    if vault_variables and charm.model.relations["vault"]:
+        try:
+            vault_client = charm.vault_relation.get_vault_client()
+        except Exception as e:
+            logger.error("Unable to initialize vault client: %s", e)
+            raise ValueError("Unable to initialize vault client. Remove relation and retry.") from e
+
+        for item in vault_variables:
+            key = item.get("key")
+            path = item.get("path")
+            try:
+                secret = vault_client.read_secret(path=path, key=key)
+            except Exception as e:
+                raise ValueError(f"Unable to read vault secret `{key}` at path `{path}`: {e}") from e
+            charm_env.update({key: secret})
+
+    return charm_env
+
+
+def parse_secrets(yaml_string):
```

Comment:
maybe we should use pydantic here?

Reply:
I've added a TODO for converting the charm to use structured config instead